subsys/bluetooth/host/smp.c

```diff
@@ -284,7 +284,19 @@ static bool sc_oobd_present;
 static bool legacy_oobd_present;
 static bool sc_supported;
 static const uint8_t *sc_public_key;
-static K_SEM_DEFINE(sc_local_pkey_ready, 0, 1);
+
+static void bt_smp_pkey_ready(const uint8_t *pkey);
+static struct {
+	struct k_mutex lock;
+	struct k_condvar condvar;
```

Comment:
I wasn't expecting to see these primitives in the solution, based on our discussion in last week's WG meeting. There was talk about using a "signal", which I assumed to be reference to `k_poll_signal`. @alwa-nordic or did you mean something else?

Reply:
Please correct me if I'm wrong, but signal won't work in this scenario as expected as we need to call [`k_poll_signal_reset`](https://github.com/zephyrproject-rtos/zephyr/blob/ece3f3b5b63ec56eb240584da9491b32dc7dcc77/include/zephyr/kernel.h#L6151-L6155) to clear the signal otherwise the lock won't work. But then it is a question: which thread should reset the signal? What we want to achieve here is when multiple threads enters this code and wait for the key, they all are released with the key (or not) and this can only be achieved with conditional variable + broadcast.

Reply:
Previously I suggested using a signal because I thought `bt_pub_key_gen` should be invoked once from `smp_init`, and then never again. I was not aware that `bt_pub_key_gen` can fail with a transient error.

Since `bt_pub_key_gen` can fail spuriously and has to be retried, a `k_poll_signal` is not sufficient since that would record the failure and make it permanent. @PavelVPV's solution is more correct.

Reply:
@PavelVPV is there a distinct benefit (or perhaps even a requirement) of using `k_mutex` instead of `k_sem`? Ever since listening to [Andy's presentation on Zephyr synchronization primitives](https://www.youtube.com/watch?v=p8OgqVQxklo&pp=ygUbYW5keSByb3NzIHplcGh5ciBwcmltaXRpdmVz), I've tried to steer clear of `k_mutex` whenever possible.

Reply:
`k_condva_wait` API? https://github.com/zephyrproject-rtos/zephyr/blob/50c21f159143e18c8a05cbb946a5e7610033355f/include/zephyr/kernel.h#L3198-L3199

Reply:
@PavelVPV thanks, I should have looked at the details. This is the first time we're introducing any kind of (AFAIK) `k_mutex` usage into the Bluetooth subsystem, so this raised some eyebrows.